### PR TITLE
Fail_all_ pending txs (including `CREATED` and `READY`)

### LINKED ISSUE #2997: Unfinished transactions not able to be canceled
## Steps to reproduce

1. You will need to be connected using Metamask.
2. Create a new action with multiple transactions e.g. any action with a description.
3. Ensure the pending transaction button appears in the header.
4. Don't sign the transaction, just reload the page.
6. Click on the pending transaction pill still in the header.
7. Notice that you will have pending transactions still, however, no way to cancel them.

Note: I went through this with Chris on a call and we decided it made sense to remove the cancel button that used to be here as it was a duplicate of the cancel option that was within the errors notice and only came up because retrying was broken.

However, it seems we were misled by the broken retry and this cancel button needs to still be there.

Discussion point: It would be good to be able to retry the transaction, but there seems to be a risk of not knowing exactly what you are retyring. Keen to hear thoughts on this.

## Expected behaviour

* You should be able to cancel transactions that were not completed, so that the pending pill is not always visible.
* In progress/incomplete transactions should show a loader with an option to cancel the transaction inline.
    - Upon clicking the “Cancel” button, the user should get a “Yes / No” confirmation.
        - Clicking “Yes” will cancel the transaction.
        - Clicking “No” will return to the previous state showing the “Cancel” button and keep the transaction as pending.
        - The cancel option should not be available if the transaction has already been executed.
        - All cancel, yes and no buttons should have a hover state of Blue/400.

[Figma link - Userhub transactions](https://www.figma.com/design/l1dOM5qiQYwF0ElvKDqqjg/Design-System---Colony-v3?node-id=5072-2755&t=rNj0waNyXDTB0pHC-4)
image

## Actual behaviour

* You are not able to cancel or retry transactions in the transactions tab that did not fail because of an error.
* You have a pending transaction button appearing in the header indefinitely.

image

image

### PR DESCRIPTION
**Pull request by chmanie**
## Description

This PR adds transactions with the status `CREATED` and `READY` to the ones that are being cleared when the user refreshes the page.

I know this is not the ideal approach (we would like to retry those transactions), but with the current transaction API that's all we can do for now.

To fix the current issues on QA and production we will have to write a migration script that fails all of these already existing transactions.

## Testing

This is quite hard to test locally, so you will have to create a transaction for your user manually:

1) Create unfinished transactions manually:

```gql
mutation CreateUnfinishedTransactions {
  tx1: createTransaction(
    input: {id: "zBtVbwKJvFhxKbJSUTbma1-createColonyForFrontend", context: NetworkClient, from: "0xb77D57F4959eAfA0339424b83FcFaf9c15407461", colonyAddress: "0x52fC5a44C7f53C42f91BB20c2C9304F046481e91", groupId: "zBtVbwKJvFhxKbJSUTbmy", group: {id: "zBtVbwKJvFhxKbJSUTbma-createColonyForFrontend", groupId: "zBtVbwKJvFhxKbJSUTbmy", index: 10, key: ""}, methodName: "createColonyForFrontend", status: CREATED, blockHash: null, blockNumber: null, createdAt: "2024-08-26T14:22:03.445Z", hash: "0xa015a493bdc50ff0488e3522ea976e8c8be3e1239f3114199f0ec35b13a34088", identifier: "0x52fC5a44C7f53C42f91BB20c2C9304F046481e91", methodContext: null, options: "{}", params: "[\"0x0000000000000000000000000000000000000000\",\"Space Credits\",\"CREDS\",18,15,\"\",\"\"]", receipt: null, title: null, titleValues: null}
  ) {
    id
  }
  tx2:createTransaction(
    input: {id: "zBtVbwKJvFhxKbJSUTbmb1-createColonyForFrontend", context: NetworkClient, from: "0xb77D57F4959eAfA0339424b83FcFaf9c15407461", colonyAddress: "0x52fC5a44C7f53C42f91BB20c2C9304F046481e91", groupId: "zBtVbwKJvFhxKbJSUTbmy", group: {id: "zBtVbwKJvFhxKbJSUTbma-createColonyForFrontend", groupId: "zBtVbwKJvFhxKbJSUTbmy", index: 10, key: ""}, methodName: "createColonyForFrontend", status: READY, blockHash: null, blockNumber: null, createdAt: "2024-08-26T14:22:03.445Z", hash: "0xa015a493bdc50ff0488e3522ea976e8c8be3e1239f3114199f0ec35b13a34088", identifier: "0x52fC5a44C7f53C42f91BB20c2C9304F046481e91", methodContext: null, options: "{}", params: "[\"0x0000000000000000000000000000000000000000\",\"Space Credits\",\"CREDS\",18,15,\"\",\"\"]", receipt: null, title: null, titleValues: null}
  ) {
    id
  }
  tx3: createTransaction(
    input: {id: "zBtVbwKJvFhxKbJSUTbmc1-createColonyForFrontend", context: NetworkClient, from: "0xb77D57F4959eAfA0339424b83FcFaf9c15407461", colonyAddress: "0x52fC5a44C7f53C42f91BB20c2C9304F046481e91", groupId: "zBtVbwKJvFhxKbJSUTbmy", group: {id: "zBtVbwKJvFhxKbJSUTbma-createColonyForFrontend", groupId: "zBtVbwKJvFhxKbJSUTbmy", index: 10, key: ""}, methodName: "createColonyForFrontend", status: PENDING, blockHash: null, blockNumber: null, createdAt: "2024-08-26T14:22:03.445Z", hash: "0xa015a493bdc50ff0488e3522ea976e8c8be3e1239f3114199f0ec35b13a34088", identifier: "0x52fC5a44C7f53C42f91BB20c2C9304F046481e91", methodContext: null, options: "{}", params: "[\"0x0000000000000000000000000000000000000000\",\"Space Credits\",\"CREDS\",18,15,\"\",\"\"]", receipt: null, title: null, titleValues: null}
  ) {
    id
  }
}
```

2. Refresh the app using the address `0xb77D57F4959eAfA0339424b83FcFaf9c15407461` (Leela) and check if all transactions are either failed or succeeded.

## Diffs

**Changes** 🏗

* Also fail transactions that are `CREATED` or `READY` on user login

Resolves #2997

### CODE REVIEW

**Review by jakubcolony — APPROVED** [review]
Approving as I can confirm the transactions were failed in the DB:
image

I was expecting to see them at the bottom of transactions in the user hub (that's how they're returned by the query), but they're not in the list:
image

Is that expected? 

----

Most likely unrelated to this PR, but twice I got into a state where no transactions would show in the user hub for my user:
image

I had to change the fetch policy of `useGetUserTransactionsQuery` in `src/state/transactionState.ts` to network-only to see transactions for my user. Can you check if you're able to reproduce?

**Comment by chmanie** [comment]
I was seeing something similar but thought it was unrelated. I'll investigate!

**Review by bassgeta — DISMISSED** [review]
Nice job fixing this, I went through the steps and only got successful transactions :+1: 
image

**Comment by chmanie** [comment]
I will do a little change for this and add a lambda to fail the transactions

**Comment by chmanie** [comment]
A LITTLE CHANGE he said

File: amplify/backend/api/colonycdapp/schema/schema.graphql [hunk 1]
```diff
@@ -973,22 +973,30 @@ type Mutation {
...
     input: UpdateContributorsWithReputationInput!
   ): Boolean @function(name: "updateContributorsWithReputation-${env}")
-
   """
   Validates the user invite code and adds the user as a colony contributor
   """
   validateUserInvite(input: ValidateUserInviteInput!): Boolean
     @function(name: "validateUserInvite-${env}")
   """
+  Initialize user (for now only cancels pending transactions)
```
Comment (chmanie) [inline amplify/backend/api/colonycdapp/schema/schema.graphql:982]: I gave this a bit of a more general name to be able to add more to this lambda if we so wish

File: amplify/backend/function/initializeUser/initializeUser-cloudformation-template.json
Comment (chmanie) [inline amplify/backend/function/initializeUser/initializeUser-cloudformation-template.json:1]: I'm not super sure if this is needed, please help, lambda-people

File: docker/colony-cdapp-dev-env-auth-proxy [hunk 2]
```diff
@@ -1,6 +1,6 @@
 FROM colony-cdapp-dev-env/base:latest
 
-ENV AUTH_PROXY_HASH=f1d5e35271358946ec65e9a0f7d27fa66c2b878a
+ENV AUTH_PROXY_HASH=b83e72fd73bc676fc78872d2a646c140e521c0b3
```
Comment (chmanie) [inline docker/colony-cdapp-dev-env-auth-proxy:3]: This needs to be adjusted once https://github.com/JoinColony/colonyCDappAuthProxy/pull/25 and https://github.com/JoinColony/colonyCDappAuthProxy/pull/26/files are merged

File: src/state/transactionState.ts [hunk 3]
```diff
@@ -527,45 +522,28 @@ export const failPendingTransactions = async () => {
...
-  } while (nextToken);
-
-  if (!promises) {
-    return;
-  }
-
-  await Promise.allSettled(promises);
+      update: (cache, { data: result }) => {
+        if (result?.initializeUser) {
+          result.initializeUser.failedTransactions.forEach((failedTx) => {
```
Comment (chmanie) [inline src/state/transactionState.ts:535]: This adjusts the Apollo cache accordingly for all failed transactions.

**Review by jakubcolony — CHANGES_REQUESTED** [review]
Changes look sensible, however I can't seem to be able to connect my wallet due to the mutation failing. I tried it a bunch of times and it worked once out of 10 attempts so maybe there's a race condition?


https://github.com/user-attachments/assets/90c35c3b-4f9e-4a0a-9ce4-47925d1bd240

File: amplify/backend/function/initializeUser/src/index.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+      promises,
+      result.data.getTransactionsByUser?.items.filter((i) => (!!i)).map((tx) => {
+        return graphqlRequest(
+          failTransaction,
+          { id: tx.id },
+          graphqlURL,
+          apiKey,
+        );
+      }),
+    );
```
[the remark this answers is not in the document]
Comment (chmanie) [reply amplify/backend/function/initializeUser/src/index.js:73]: I have absolutely no idea :D

What was I thinking? I'll change that ;)

File: amplify/backend/function/initializeUser/src/utils.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  // Colony tokens have the `locked()` method. We assume that when it exists on
+  // the contract we have a ColonyToken 🦆. This might not be true though, so can't rely
+  // on this 100% when trying to call contract methods
+  try {
+    await client.locked();
+    return TokenType.COLONY;
+  } catch {
+    return TokenType.ERC20;
+  }
+};
```
[the remark this answers is not in the document]
Comment (chmanie) [reply amplify/backend/function/initializeUser/src/utils.js:47]: Good catch that must have been a c&p thing

File: amplify/backend/function/initializeUser/src/utils.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
+const fetch = require('cross-fetch');
+
+const TokenType = {
+  COLONY: 'COLONY',
+  ERC20: 'ERC20',
+};
+
+const graphqlRequest = async (queryOrMutation, variables, url, authKey) => {
```
Comment (chmanie) [inline amplify/backend/function/initializeUser/src/utils.js:8]: And how did this ever work without being exported??

**Comment by chmanie** [comment]
> I tried it a bunch of times and it worked once out of 10 attempts so maybe there's a race condition?

You were indeed right about the race condition. I must have been in a nebulous state of mind when I wrote this code. Please try again now :)

**Review by jakubcolony — CHANGES_REQUESTED** [review]
After the changes I'm still unable to connect my wallet as the mutation keeps returning 403. I think you may need to always allow it in the auth proxy so that it can make it to the lambda.


https://github.com/user-attachments/assets/ab72aa13-975d-4421-900f-bcd61cf76ea4

**Review by jakubcolony — COMMENTED** [review]
After I changed the auth proxy hash to point at [colonyCDappAuthProxy#29](https://github.com/JoinColony/colonyCDappAuthProxy/pull/29) I was able to connect wallet so that issue is resolved 👍 

As you can see on the recording, I ran the mutation with only IDs and createdAt fields changed and confirmed the transactions had been failed, however the `failedTransactions` array from the lambda is empty. I was expecting it to contain the 3 transactions I created.

Additionally, this may not be relevant to this PR, but want to confirm if it's a genuine issue. The transactions I created are not showing in the user hub transactions list despite being returned by the query. Their createdAt field is set to the future so they should be on the top of the list. 


https://github.com/user-attachments/assets/82e1419f-faea-4b8a-aa20-03afe3568723


https://github.com/user-attachments/assets/82e1419f-faea-4b8a-aa20-03afe3568723

**Comment by chmanie** [comment]
Alright, I fixed the issue that the `failedTransactions` wouldn't return anything and bumped the auth proxy hash to the current master which should work.

> Additionally, this may not be relevant to this PR, but want to confirm if it's a genuine issue. The transactions I created are not showing in the user hub transactions list despite being returned by the query. Their createdAt field is set to the future so they should be on the top of the list.

Just by looking at it I can't figure out why they are not shown either, nevertheless, this is a bit of a contrived case. Try to create transactions that actually contain all required fields and see if they're not showing, still. Otherwise feel free to open an issue for this.

**Review by jakubcolony — APPROVED** [review]
UI initiated transactions are showing in the list correctly, so there's probably nothing to worry about.

This time I got the 3 transactions in the response array:
image

Nice work @chmanie 💯

**Review by iamsamgibbs — APPROVED** [review]
Nice changes @chmanie ! Having this in a lambda feels quite a bit neater than what was here before. Tested and can see the transactions getting failed:

Screenshot 2024-09-19 at 09 36 25

I also couldn't see them in the UI. I wonder if it's possible to create a testing scenario via the UI where a transaction is deliberately left in a created / pending / ready state and checking if it properly failed when refreshing?

**Pull request merged**